Mathlib/Data/ZMod/IntUnitsPower.lean

```diff
@@ -9,56 +9,107 @@ import Mathlib.Data.Int.Order.Units
+# The power operator on `ℤˣ` by `ZMod 2`, `ℕ`, and `ℤ`
 
 See also the related `negOnePow`.
 
 ## TODO
 
 * Generalize this to `Pow G (Zmod n)` where `orderOf g = n`.
-* Abstract this with a `LawfulPow` typeclass such that we can reuse the same lemmas for `z₂pow`,
-  `npow`, and `zpow`.
+
+## Implementation notes
+
+In future, we could consider a `LawfulPower M R` typeclass; but we can save ourselves a lot of work
+by using `Module R (Additive M)` in its place, especially since this already has instances for
+`R = ℕ` and `R = ℤ`.
 -/
 
 local macro_rules | `($x ^ $y) => `(HPow.hPow $x $y) -- Porting note: See issue lean4#2220
 
-/-- There is a canonical power operation by `ℤˣ` on `ZMod 2`.
+instance : SMul (ZMod 2) (Additive ℤˣ) where
+  smul z au := .ofMul <| Additive.toMul au ^ z.val
+
+lemma ZMod.smul_units_def (z : ZMod 2) (au : Additive ℤˣ) :
+    z • au = z.val • au := rfl
+
+lemma ZMod.natCast_smul_units (n : ℕ) (au : Additive ℤˣ) : (n : ZMod 2) • au = n • au :=
+  (Int.units_pow_eq_pow_mod_two au n).symm
+
+/-- This is an indirect way of saying that `ℤˣ` has a power operation by `ZMod 2`. -/
+instance : Module (ZMod 2) (Additive ℤˣ) where
+  smul z au := .ofMul <| Additive.toMul au ^ z.val
+  one_smul au := Additive.toMul.injective <| pow_one _
+  mul_smul z₁ z₂ au := Additive.toMul.injective <| by
+    dsimp only [ZMod.smul_units_def, toMul_nsmul]
+    rw [←pow_mul, ZMod.val_mul, ←Int.units_pow_eq_pow_mod_two, mul_comm]
+  smul_zero z := Additive.toMul.injective <| one_pow _
+  smul_add z au₁ au₂ := Additive.toMul.injective <| mul_pow _ _ _
+  add_smul z₁ z₂ au := Additive.toMul.injective <| by
+    dsimp only [ZMod.smul_units_def, toMul_nsmul, toMul_add]
+    rw [←pow_add, ZMod.val_add, ←Int.units_pow_eq_pow_mod_two]
+  zero_smul au := Additive.toMul.injective <| pow_zero (Additive.toMul au)
+
+section CommSemiring
+variable {R : Type*} [CommSemiring R] [Module R (Additive ℤˣ)]
+
+/-- There is a canonical power operation on `ℤˣ` by `R` if `Additive ℤˣ` is an `R`-module.
+
+In lemma names, this operations is called `uzpow` to match `zpow`.
+
+Notably this is satisfied by `R ∈ {ℕ, ℤ, ZMod 2}`. -/
+instance Int.instUnitsPow : Pow ℤˣ R where
+  pow u r := Additive.toMul (r • Additive.ofMul u)
+
+-- The above instance forms no typeclass diamonds with the standard power operators
+example : Int.instUnitsPow = Monoid.Pow := rfl
+example : Int.instUnitsPow = DivInvMonoid.Pow := rfl
+
+@[simp] lemma ofMul_uzpow (u : ℤˣ) (r : R) : Additive.ofMul (u ^ r) = r • Additive.ofMul u := rfl
+
+@[simp] lemma toMul_uzpow (u : Additive ℤˣ) (r : R) :
+  Additive.toMul (r • u) = Additive.toMul u ^ r := rfl
+
+@[norm_cast] lemma uzpow_natCast (u : ℤˣ) (n : ℕ) : u ^ (n : R) = u ^ n := by
+  change Additive.toMul ((n : R) • Additive.ofMul u) = _
+  rw [←nsmul_eq_smul_cast, toMul_nsmul, toMul_ofMul]
+
+lemma uzpow_ofNat (s : ℤˣ) (n : ℕ) [n.AtLeastTwo] :
+    s ^ (no_index (OfNat.ofNat n : R)) = s ^ (no_index (OfNat.ofNat n : ℕ)) :=
```

Comment:
Can you add a note explaining the reason for the `noindex`? Or, if it makes sense, create a library note in an appropriate location and add a link to it here.

Reply:
I added a reference to the existing library note